src/Raven.Server/Utils/CpuUsageExtensionPoint.cs

```diff
@@ -0,0 +1,263 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Timers;
+using Raven.Client.Extensions;
+using Raven.Server.NotificationCenter.Notifications;
+using Raven.Server.NotificationCenter.Notifications.Details;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Logging;
+using Timer = System.Timers.Timer;
+
+namespace Raven.Server.Utils
+{
+    public class CpuUsageExtensionPoint : IDisposable
+    {
+        private readonly JsonContextPool _contextPool;
+        private readonly Logger _logger;
+        private readonly NotificationCenter.NotificationCenter _notificationCenter;
+        private readonly ProcessStartInfo _startInfo;
+        private readonly TimeSpan _timeout = TimeSpan.FromSeconds(10);
+
+        private Process _process;
+        private bool _didRestart;
+
+        private ExtensionPointRawData _data = new ExtensionPointRawData();
+        public ExtensionPointRawData BadData = new ExtensionPointRawData
+        {
+            ActiveCores = -1,
+            MachineCpuUsage = -1
+        };
+
+        public ExtensionPointRawData Data
+        {
+            get
+            {
+                if (IsDisposed && _logger.IsOperationsEnabled)
+                {
+                    _logger.Operations($"Try to get data from {nameof(CpuUsageExtensionPoint)} which was disposed");
+                }
+                return _data;
+            }
+        }
+
+        public bool IsDisposed { get; private set; }
+
+
+        public CpuUsageExtensionPoint(
+            JsonContextPool contextPool,
+            string exec,
+            string args,
+            Logger logger,
+            NotificationCenter.NotificationCenter notificationCenter)
+        {
+            _contextPool = contextPool;
+            _logger = logger;
+            _notificationCenter = notificationCenter;
+            _startInfo = new ProcessStartInfo
+            {
+                FileName = exec,
+                Arguments = args,
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                CreateNoWindow = true
+            };
+
+            InitNewProcessHandle();
+        }
+
+        public void Start()
+        {
+            try
+            {
+                _process.Start();
+            }
+            catch (Exception e)
+            {
+                NotifyWarning("Could not start cpu usage extension point process", e);
+
+                Dispose();
```

Comment:
Need to rethrow here?

Reply:
@ppekrol As I understand it the error handling should be soft so I just notify there is a problem. What do you think? how the error should be handled?

Reply:
Then you probably want to return here, not 'ReadProcess' ? (check @maximburyak comment)

Reply:
In this case, the origin methods to calculate CPU usage give wrong values so if the process didn't start or if it failed while running I still want to read the data from the disposed object to get error value from it. Another option I can think of is to switch the `CpuUsageExtensionPoint` to "Failed object" that will do this work and use it is I can't start. Is it make sense?

Reply:
If process died, then you want to `ReadErrors`, not the normal output I think?

Reply:
There is a task that responsible to read errors, notify them and kill the process if needed. But after that or if the process killed/existed by another reason when for instance the Dashboard requests for CPU usage it will get negative values to indicate the problem